crates/ironrdp-web/src/audio.rs

```diff
@@ -0,0 +1,773 @@
+                Ok(buffer) => {
+                    if let Err(e) = self.audio_queue.enqueue_audio(buffer, &self.gain_node) {
+                        error!("Failed to enqueue audio: {:?}", e);
+                    } else {
+                        info!("Successfully processed audio format: {:?}", format.format);
+                    }
+                }
+                Err(e) => {
+                    warn!("Failed to create audio buffer for format {:?}: {:?}", format.format, e);
+                }
+            }
+        } else {
+            // Context not ready, buffer the audio data
+            if self.pending_audio_data.len() < 10 {
+                // Limit buffer size
+                self.pending_audio_data.push_back((data.to_vec(), format));
+                info!(
+                    "Audio data buffered (context not ready), queue size: {}",
+                    self.pending_audio_data.len()
+                );
+            } else {
+                warn!("Audio buffer full, dropping oldest data (consider user gesture to activate audio)");
+                self.pending_audio_data.pop_front();
+                self.pending_audio_data.push_back((data.to_vec(), format));
+            }
+        }
+    }
+
+    fn set_volume(&mut self, volume: VolumePdu) {
+        info!(
+            "Setting volume: left={}, right={} (gain: {:.2})",
+            volume.volume_left,
+            volume.volume_right,
+            (f32::from(volume.volume_left) + f32::from(volume.volume_right)) / (2.0 * 65535.0)
+        );
+
+        // Convert RDP volume (0-65535) to Web Audio gain (0.0-1.0)
+        // For simplicity, use average of left and right channels
+        let left_gain = f32::from(volume.volume_left) / 65535.0;
+        let right_gain = f32::from(volume.volume_right) / 65535.0;
+        self.volume = (left_gain + right_gain) / 2.0;
+
+        if let Err(_e) = self.apply_volume() {
+            error!("Failed to apply volume");
+        }
+    }
+
+    fn set_pitch(&mut self, pitch: PitchPdu) {
+        info!(
+            "Setting pitch: {} (note: pitch control not implemented for web)",
+            pitch.pitch
+        );
+
+        // Store pitch value but don't implement pitch shifting for now
+        // Web Audio API pitch shifting would require more complex processing
+        self.pitch = pitch.pitch as f32 / 65535.0;
+    }
+
+    fn close(&mut self) {
+        info!("Closing audio backend");
+        self.is_active = false;
+
+        // Suspend audio context to free resources
+        if let Err(e) = self.audio_context.suspend() {
+            warn!("Failed to suspend audio context: {e:?}");
+        }
+
+        // Audio queue cleanup (buffers are automatically cleaned up when context suspends)
+        debug!("Audio backend closed, context suspended");
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_convert_pcm_to_float_valid_16bit() {
```

Comment:
note: Unit tests are not run by default (see the Cargo.toml), because:
- https://github.com/Devolutions/IronRDP/blob/master/ARCHITECTURE.md#testing
- https://github.com/Devolutions/IronRDP/blob/39566bff58b07637bc89d78570e043befb5123a7/crates/ironrdp-testsuite-core/tests/main.rs#L3-L12

But you don’t have to care now, I can look into moving that on my side, in a follow up PR.